jlm/rvsdg/NodeNormalization.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_NODENORMALIZATION_HPP
+#define JLM_RVSDG_NODENORMALIZATION_HPP
+
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+class output;
+
+/**
+ * \brief Normalizes the operands of an operation.
+ *
+ * The usage expectation is that IsApplicable() is invoked to determine whether operands can be
+ * normalized BEFORE ApplyNormalization() is invoked.
+ *
+ * @tparam TOperation The operation for which to normalize the operands.
+ */
```

Comment:
While generally I do not object to explicit interfaces, but given the suggestion below this is a somewhat verbose way of writing:

template<class TOperation> using NodeNormalization = std::function< std::optional<...>(const TOperation&, const std::vector<...>)>;

Reply:
I would like the normalizations to be able to carry state. Going forward, I would like to implement  the following normalizations:

1. A normalization that takes a vector of normalizations and applies them in sequence
2. A normalization that takes a single normalization (which can be a vector normalization) and applies it until a fixpoint is reached.
3. A normalization that takes a normalization and applies it n times. This can maybe combined with point 2, not sure yet.

Either way, I would like to have the option for them to carry state.

Reply:
but in the interface description you already explicitly rule state that they need to be reentrant -- which means they will not be able to carry state across multiple calls

neither fixpoint iteration nor chain of other normalizers requires state across calls

Reply:
What do you mean by "state across calls"? State that is instantiated in IsApplicable() and used in ApplyNormalization()? Or state between multiple invocations of IsApplicable()/ApplyNormalization()?